### PR TITLE
Improved variable names in MSE

### PR DESCRIPTION
**Pull request by RishabhGarg108**
In accordance with #2005 ,
After giving a careful thought and looking at the code, I figured out that it is not just the `input` variable that is somewhat misleading, but the `output` variables is also a little imprecise. 
To, explain my thoughts, consider a neural network or in general any ML `model`. The `model` takes some `input` and spits out some `predictions`. Then those `predictions` along with the `target labels` are fed into a `loss function` that returns `loss` which measures the goodness of `prediction`. (Being precise with the choice of words here)

Now, in the current implementation of loss functions, we have an `input` variable, that is actually the `prediction` and `output` is actually the calculated `loss` value.

Although all the variable names in the current implementation are correct, but they are somewhat generic in the sense that by nature, every function takes some input and spits output. This sometimes misleads contributors to do wrong implementations. So, as far as the loss functions are concerned, we can be a bit more precise with these names and choose them so that they are intuitive, i.e., we can replace 

- input -> prediction 
- output -> loss

These are exactly the changes that I have dome in this PR. Although this confusion was mostly in `Backward()` function, but to be consistent with the naming, it had to be changed in all the functions.

In this PR, I have done it for just a single one and once these changes are accepted, I will change the remaining ones too. Also, there is no right or wrong way to tackle this issue. It is just a naming convention. So, if anyone has some better naming convention than this one, then please freely suggest here. I am open to discuss it :)

Also, sorry for being too much verbose here, but I felt that I needed to explain my thought process here :D

### CODE REVIEW

**Comment by RishabhGarg108** [comment]
@rcurtin , can you please take a look at this one.

File: src/mlpack/methods/ann/loss_functions/mean_squared_error.hpp [hunk 1]
```diff
@@ -42,24 +42,26 @@ class MeanSquaredError
   /**
    * Computes the mean squared error function.
    *
-   * @param input Input data used for evaluating the specified function.
+   * @param prediction Predictions used for evaluating the specified loss
+   * function.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/mean_squared_error.hpp:46]: This should be tabbed twice (4 spaces), see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping for more details.

File: src/mlpack/methods/ann/loss_functions/mean_squared_error.hpp [hunk 2]
```diff
@@ -42,24 +42,26 @@ class MeanSquaredError
...
    * @param target The target vector.
    */
-  template<typename InputType, typename TargetType>
-  typename InputType::elem_type Forward(const InputType& input,
+  template<typename PredictionType, typename TargetType>
+  typename PredictionType::elem_type Forward(const PredictionType& prediction,
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/mean_squared_error.hpp:50]: Looks like the parameter below (`target`) don't align with the first one.

**Review by zoq — COMMENTED** [review]
I think changing the naming scheme makes sense.

**Comment by RishabhGarg108** [comment]
Hey @zoq , I made the requested changes, can you take a look again. Thanks :)

**Comment by RishabhGarg108** [comment]
> I think changing the naming scheme makes sense.

That's great! Then I will do the similar changes to all the loss functions. Till then please do not merge this one :)

**Comment by rcurtin** [comment]
> Also, sorry for being too much verbose here, but I felt that I needed to explain my thought process here :D

I haven't had a chance to look at the PR yet but no need to apologize for this!  Having the whole reasoning here makes it a lot easier to review and understand why things have been changed. :+1:

**Comment by RishabhGarg108** [comment]
> I haven't had a chance to look at the PR yet but no need to apologize for this! Having the whole reasoning here makes it a lot easier to review and understand why things have been changed.

Okay :+1: . Have a look at it when you get some time. I will wait for your take on this before moving forward with this one :)

**Review by zoq — APPROVED** [review]
Thanks, looks great, thanks for the contribution.

**Comment by RishabhGarg108** [comment]
Always happy to contribute :D
Thanks @zoq , for taking time to review this.
Please don't merge this one. I will make similar changes to other loss functions in this one. Thanks!

**Comment by zoq** [comment]
Let's open a new PR for further changes, makes the review easier.

**Comment by RishabhGarg108** [comment]
> Let's open a new PR for further changes, makes the review easier.

Okay:+1: That's fine.

**Review by birm — APPROVED** [review]
You're right; "prediction" _is_ more descriptive than "input" here.

**Comment by shrit** [comment]
@zoq I think build errors are not related, can we merge this?

**Comment by zoq** [comment]
> @zoq I think build errors are not related, can we merge this?

Yes, let's merge this one.

**Comment by zoq** [comment]
@RishabhGarg108 thanks for the contribution.

**Comment by RishabhGarg108** [comment]
@zoq @birm thanks for taking time to review this :)
Can you also take a look at #2724 . This one is in continuation to this one and I have essentially applied the same changes (that I have done here) to all the loss functions. Thanks!
